### PR TITLE
monitor ssh connection live-ness

### LINKED ISSUE #12139: local-exec and remote-exec resource appears stalled while Terraform keeps waiting for them and output from script is not displayed [ was Getting the remote-exec provisioner to detect SSH connectivity issues (with or without bastion_host)]
Hi.
It seems like my 'remote-exec' resource continue running after the connection to my SSH bastion (and/or final target) drops. I haven't been able to identify the cause of the disconnection. but clearly Terraform 0.8.7 does not detect it. I was looking at the 'connection' parameters to see if I could enable a keep-alive as workaround but it does not seem supported. 

This seems highly intermittent and potentially related to some instability with our local network at the moment but I'd rather have a fast failure than a stuck resource creation process.

** UPDATE 2017-03-22 **
I'm getting the same result using local-exec that SSH to the target. 
I'm getting the same result regardless of the use of an SSH bastion or not.
At the same time, I'm getting SSH disconnection when connected to the target using a separate shell.

Either something is re-spawning the SSH server on my server or something in the AWS network gets reconfigured and/the connection teared down as I suspected initially. So I understand one could argue that this is NOT a Terraform bug, per se. However, is must be possible to mitigate this problem.

- enabling TCPKeepAlive on the bastion connection would not be very useful on its own considering that the default keep alive interval on Linux is 2 hours, I didn't check if the value was overridden by the Golang client code yet.  
- similarly, setting up ServerAliveInterval ServerAliveCountMax options in client could help 'fail fast' in case of network disruption and prevent long delays.

### Terraform Version
0.8.8

### Affected Resource(s)
Please list the resources as a list, for example:
- null-resource with remote-exec using ssh_bastion

### Terraform Configuration Files

resource "null_resource" "prep_install" {
  connection {
    user                = "${var.ssh_user}"
    private_key         = "${var.ssh_private_key}"
    timeout             = "10m"                               
    host                = "${element(aws_instance.instance.*.private_ip, count.index)}"
    bastion_host        = "${var.bastion_host}"
    bastion_user        = "${var.bastion_user}"
    bastion_private_key = "${var.bastion_private_key}"
  }

  provisioner "file" {
    source      = "${path.module}/scripts/"
    destination = "/tmp/"
  }

  provisioner "remote-exec" {
    inline = [
      "set -eu",
      "sudo chmod +x /tmp/*.sh",
      "sudo /tmp/run.sh",
    ]
  }
}

module.my.null_resource.prep_install.0 (remote-exec): <running and producing output>
module.my.null_resource.prep_install.0: Still creating... (1m50s elapsed)
module.my.null_resource.prep_install.0: Still creating... (2m0s elapsed)
...

### Expected Behavior
30 seconds after connection lost for instance. (no output or keepalive from remote-exec), the resource creation would fail.
or
SSH reconnection would be attempted and the remote-exec resumed somehow

### Actual Behavior
the remote-exec seem to have completed while being disconnected from the session
the resource creation failed after the specified timeout

### Steps to Reproduce
unable to reproduce yet

### Important Factoids
AWS VPC

### PR DESCRIPTION
**Pull request by jbardin**
An unresponsive server or dead tcp connection can hang a provisioner indefinitely.
Monitor the ssh connection status, and abort if the server becomes unresponsive for more than 120s.

Unfortunately the connection `timeout` parameter is already taken to mean "retry timeout", and is often set to an inappropriate value to use for monitoring the connection status. Rather than providing another similarly named option for users, we'll set a conservative timeout of 120s (somewhat arbitrarily chosen to match the arbitrarily chosen default TCP MSL value).

Fixes #12139

### CODE REVIEW

File: communicator/ssh/communicator_test.go [hunk 1]
```diff
@@ -100,16 +100,19 @@ func newMockLineServer(t *testing.T, signer ssh.Signer, pubKey string) string {
 
 			go func(in <-chan *ssh.Request) {
 				for req := range in {
+					// since this channel's requests are serviced serially,
```
Comment (pselle) [inline communicator/ssh/communicator_test.go]: Nice comment, and nicely done

**Comment by jbardin** [comment]
The `ssh` package is part of the `golang.org/x/crypto` module, so they come as one unit, and it's also generally a good idea to keep the crypto stuff up to date. The `golang.org/x/` packages tend to be developed in concert with the std lib, and I like to update them along with Go releases (though they don't have specific releases of their own).

**Comment by apparentlymart** [comment]
I agree that keeping these dependencies up-to-date is good, but I'd say let's also review the changelog for the library to see if there are any changes that would effect emergent behavior for our end-users.

The SSH features in particular have been a source of minor churn in the past, I seem to recall, because the support for various SSH features in the upstream library has shifted over time. In this case, just from skimming the diff it looks like e.g. the library may now have GSSAPI (Kerberos) support, which would be great to declare in the changelog if we can verify that is so. (Getting a Kerberos-enabled sshd running to test it is likely bothersome, though. :confounded: )

**Comment by jbardin** [comment]
There weren't many changes:

```
$ git log --oneline a29dc8fdc734..4def268fd1a4
4def268 (HEAD -> master, origin/master, origin/HEAD) ssh: skip testHandshakeErrorHandlingN on js/wasm
cc06ce4 acme: send User-Agent and add Client.UserAgent
ea8f1a3 ed25519: turn into a wrapper for crypto/ed25519 beginning with Go 1.13
57b3e21 ssh: add diffie-hellman-group-exchange-sha256
5c40567 internal/chacha20: fix variable naming
f99c8df poly1305: improve performance with asm for ppc64le
20be4c3 internal/chacha20: improve performance for ppc64le
22d7a77 sha3: fix bug in cSHAKE Clone()
cbcb750 ssh/gss: support kerberos authentication for ssh server and client
e1dfcc5 openpgp: replace "currentTime" with "creationTime" as appropriate
```

The diffie-hellman-group-exchange-sha256 kex was primarily what I wanted to keep compatibility as broad as possible. The gssapi support might be nice to have in the future, but not something we want to try and implement right now.

**Comment by apparentlymart** [comment]
If upgrading doesn't automatically introduce any support for GSSAPI without other changes on our end then that's fine. I just worry about new features we don't expect being introduced and becoming compatibility constraints even though we didn't get a chance to design how they interact with Terraform.

The intent of my previous comment was just to ask that we make sure we know about any new features this might introduce automatically, not to make any special effort to opt into new features explicitly.

**Comment by jbardin** [comment]
Gotcha. Yes, the only new code automatically used will be the new key exchange. The auth methods need to be explicitly configured.

**Comment by apparentlymart** [comment]
Great! In that case, it sounds like just mentioning that we now support that key exchange method in the changelog will do the job here.

**Pull request merged**